ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java

```diff
@@ -1527,8 +1527,23 @@ public void truncateTable(String dbDotTableName, Map<String, String> partSpec, L
       }
 
       // TODO: APIs with catalog names
-      List<String> partNames = ((null == partSpec)
-        ? null : getPartitionNames(table.getDbName(), table.getTableName(), partSpec, (short) -1));
+      List<String> partNames = null;
+      if (table.getStorageHandler() != null && table.getStorageHandler().alwaysUnpartitioned() && partSpec != null) {
+        StringBuilder partName = null;
+        for (Map.Entry<String, String> entry : partSpec.entrySet()) {
+          if (partName == null) {
+            partName = new StringBuilder(String.format("%s=%s", entry.getKey(), entry.getValue()));
```

Comment:
do the initialization during var declaration, if spec is empty -> list would be empty

Reply:
This is no more required. `getPartitionsByFilter()` should get the list of partitions now.